api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -1016,6 +1020,19 @@ message Cluster {
   // [#extension-category: envoy.network.dns_resolver]
   core.v3.TypedExtensionConfig typed_dns_resolver_config = 55;
 
+  // [#not-implemented-hide:]
+  // Configures on-demand DNS resolution configuration. Note that this field controls the behavior of the cluster,
+  // not the DNS resolver itself (use
+  // :ref:`typed_dns_resolver_config <envoy_v3_api_field_config.cluster.v3.Cluster.typed_dns_resolver_config>`
+  // for those use-cases). At present there are 2 modes: DISABLED and DELAYED. DISABLED mode persists the background,
+  // asynchronous DNS resolution behavior that, at time of writing, is the default for STRICT_DNS and LOGICAL_DNS clusters.
+  // DELAYED mode will delay async DNS resolution until the first connection is made to the cluster in question.
+  // Once DNS resolution is complete, async resolution will continue periodically in the background as usual. Note that the latency of
+  // the DNS resolution will affect the first connection's RTT. Consequently, failure to resolve DNS before the connect timeout
+  // will result in a connection reset. This mode is useful for scenarios where large numbers of clusters must be sent to Envoy for config
+  // practicality reasons, but traffic distribution is not uniform across all clusters.
+  DelayedDns delayed_dns = 58;
```

Comment:
Yeah that approach sounds reasonable to me! Is there somewhere I can read about the difference between extensions vs main protos and the benefit of keeping these things in extensions?

Reply:
I don't know that we've explicitly written it down anywhere, but the main impetus is just to keep the API tractable.  We can't keep adding every one-off knob that everyone wants as first-class fields, because the API will quickly become too complex for anyone to understand.  Instead, we want to have a very simple overall structure with individual features in extension points, so that people that don't need a particular extension never need to see or care about the fields related to that extension.

Reply:
Ack makes perfect sense. I'll keep this PR parked and create a new one like extensions/clusters/dns.proto